src/components/__tests__/EmailForm-test.tsx

```diff
@@ -0,0 +1,144 @@
+
+    let hiddenInput = wrapper.find("[type='hidden']");
+    expect(hiddenInput.length).to.equal(1);
+    expect(hiddenInput.prop("name")).to.equal("uuid");
+    expect(hiddenInput.prop("value")).to.equal("UUID1");
+
+    let button = wrapper.find("button");
+    expect(button.length).to.equal(1);
+  });
+
+  it("displays the correct button content", () => {
+    let button = wrapper.find("button");
+    expect(button.text()).to.contain("Send Validation Email");
+    expect(button.prop("disabled")).not.to.be.true;
+
+    let library = wrapper.prop("library");
+    let validated = Object.assign(library.urls_and_contact, { "validated": "validation time" });
+    wrapper.setProps({ library: Object.assign(library, { urls_and_contact: validated })});
+    button = wrapper.find("button");
+    expect(button.text()).to.contain("Send Validation Email");
+    expect(button.prop("disabled")).not.to.be.true;
+
+    let noEmail = Object.assign(library.urls_and_contact, { contact_email: null });
+    wrapper.setProps({ library: Object.assign(library, { urls_and_contact: noEmail })});
+    button = wrapper.find("button");
+    expect(button.text()).to.equal("No email address configured");
+    expect(button.prop("disabled")).to.be.true;
+  });
+
+  it("displays the correct button icon", () => {
+    let icon = wrapper.find("button").find("svg");
+    expect(icon.length).to.equal(0);
+
+    wrapper.setState({ sent: true });
+
+    icon = wrapper.find("button").find("svg");
+    expect(icon.length).to.equal(1);
+    expect(icon.hasClass("check-solo-icon")).to.be.true;
+  });
+
+  it("displays an info message", () => {
+    let info = wrapper.find(".alert-info");
+    expect(info.length).to.equal(0);
+
+    let library = wrapper.prop("library");
+    let validated = Object.assign(library.urls_and_contact, { "validated": "validation time" });
+    wrapper.setProps({ library: Object.assign(library, { urls_and_contact: validated })});
+
+    info = wrapper.find(".alert-info");
+    expect(info.length).to.equal(1);
+    expect(info.text()).to.equal("Already validated");
+
+    // Hide the info message if there's a success message or an error message to display instead
+    wrapper.setState({ sent: true });
+    info = wrapper.find(".alert-info");
+    expect(info.length).to.equal(0);
+
+    wrapper.setProps({ error: { response: "Failed to send email" } });
+    info = wrapper.find(".alert-info");
+    expect(info.length).to.equal(0);
+  });
+
+  it("submits on click", async () => {
+    expect(wrapper.state()["sent"]).to.be.false;
+    wrapper.find("button").simulate("click");
+    expect(email.callCount).to.equal(1);
+    expect(email.args[0][0].get("uuid")).to.equal("UUID1");
+
+    const pause = (): Promise<void> => {
+      return new Promise<void>(resolve => setTimeout(resolve, 0));
+    };
+    await pause();
+
+    expect(wrapper.state()["sent"]).to.be.true;
+  });
+
+  it("displays a success message", () => {
```

Comment:
If you set `sent`'s state to true and also pass in the error prop, both the success and danger messages will show up at the same time. So might be good to check for that case in the component.

Reply:
I will go ahead and put in a safeguard just in case, but I don't see any way for this situation to arise in real life.  (See above: if there's an error prop, `sent` never gets set to true.)

Reply:
Can you think of any situation (e.g. if there were multiple fieldsets in the same form) in which, realistically, we would want a form to simultaneously display an error message AND a success message?  (I'm considering putting the safeguard into the `Form` component--making it so that it can only show one at a time.)

Reply:
I definitely do not see any case were we would want a form to display both, but it seemed like the logic would've caused both to show up. But if the `setState` function does not get called after the async function, then it's fine.